File: src/java/foundgine-core/src/main/java/com/foundgine/core/semantic/metadata/FieldMetadata.java
package com.foundgine.core.semantic.metadata;

import com.foundgine.core.abstractions.FieldId;
import java.util.List;

public record FieldMetadata(FieldId id, String name, Class<?> clrType, ColumnReference column, String dimension,
		boolean indexed, List<AliasDeclaration> aliases) {
	public FieldMetadata(FieldId id, String name, Class<?> clrType) {
		this(id, name, clrType, null, null, false, null);
	}

	public FieldMetadata(FieldId id, String name, Class<?> clrType, ColumnReference column) {
		this(id, name, clrType, column, null, false, null);
	}

	/**
	 * Returns declared aliases, matching the C# metadata's null-as-empty semantics.
	 */
	public List<AliasDeclaration> effectiveAliases() {
		return aliases == null ? List.of() : List.copyOf(aliases);
	}
}

File: src/csharp/Foundgine.Core/Semantic/Metadata/FieldMetadata.cs
using Foundgine.Core.Abstractions;

namespace Foundgine.Core.Semantic.Metadata;

/// <summary>
/// Static field metadata, including an optional provider-neutral storage
/// column reference. Providers decide how to translate that reference.
/// </summary>
public sealed record FieldMetadata(
    FieldId Id,
    string Name,
    Type ClrType,
    ColumnReference? Column = null,
    // The semantic dimension this field represents (e.g. "tenant", "country",
    // "category"), or null when the field is a plain data value with no
    // declared semantic role. See FoundgineSemanticDimensionAttribute.
    string? Dimension = null,
    // Hints that this field is backed by (or should be backed by) a storage
    // index, for use by query planners choosing access paths.
    bool IsIndexed = false,
    IReadOnlyList<AliasDeclaration>? Aliases = null);